### PR TITLE
Allow SITL on HW to drive actuators

### PR DESCRIPTION
**Pull request by joshanne**
Adds a parameter for allowing relays to be driven by SIM on HW.
This reflects similar changes that allowed driving servo channels with SIM on HW.

### CODE REVIEW

**Comment by joshanne** [comment]
@peterbarker @Hwurzburg @IamPete1 
Looks like there was an original PR that explicitly disabled Relays for SIM on HW. As mentioned, I for one have been making extensive use of this setup for integration testing on a "flight vehicle".

I add this flag as a kind of interlock, the user really has to know what they're doing when they use it, because it is interfacing to real actuators.

**Comment by peterbarker** [comment]
Interesting.  We originally also blocked servo outputs, but we came back and added a mask of servo outputs which can go out under sim-on-hardware.  Do you think we should likewise add a mask here instead of having this all-or-nothing approach?

**Comment by joshanne** [comment]
@peterbarker implemented and tested on hardware. Works as expected (after I remember to set the parameter! :stuck_out_tongue:  )

File: libraries/SITL/SITL.cpp [hunk 1]
```diff
@@ -911,6 +911,11 @@ const AP_Param::GroupInfo SIM::var_ins[] = {
     AP_GROUPINFO("GYR5_BIAS",    47, SIM, gyro_bias[4], 0),
 #endif
 
+    // @Param: OH_RELY_MSK
+    // @DisplayName: SIM-on_hardware Relay Enable Mask
+    // @Description: relays which are passed through to actual hardware when running sim on actual hardware
```
Comment (Hwurzburg) [inline libraries/SITL/SITL.cpp:916]:
```suggestion
    // @Description: Allow relay output operation when running SIM-on-hardware
```

**Review by Hwurzburg — CHANGES_REQUESTED** [review]

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: libraries/AP_Relay/AP_Relay.cpp [hunk 2]
```diff
@@ -157,7 +157,9 @@ void AP_Relay::set(const uint8_t instance, const bool value)
        _last_logged_pin_states = _pin_states;
     }
 #if AP_SIM_ENABLED && (CONFIG_HAL_BOARD != HAL_BOARD_SITL)
```
[the remark this answers is not in the document]
Comment (joshanne) [reply libraries/AP_Relay/AP_Relay.cpp]: I'm happy to remove the check for `HAL_BOARD_SITL` which brings it in line with the servo outputs.

This makes the implementation the same as that of the servo outputs.

File: libraries/SITL/SITL.cpp (hunk 1 above)
Comment (joshanne) [reply libraries/SITL/SITL.cpp]: Updated

**Comment by tridge** [comment]
@joshanne nice idea!
I'll leave @peterbarker to review

**Review by Hwurzburg — APPROVED** [review]

**Comment by joshanne** [comment]
There were a handful of legitimate autotest failures in vehicles testing relay channels. This has been fixed. 

@peterbarker looks like the remaining failures are unrelated to this PR.

File: libraries/SITL/SITL.cpp [hunk 3]
```diff
@@ -47,6 +47,15 @@ extern const AP_HAL::HAL& hal;
 #endif
 #endif
 
+#if (CONFIG_HAL_BOARD != HAL_BOARD_SITL)
+// For on-hardware, set allowed relay channels to zero.
+// Requires user to change the param to allow hadware access.
+#define SIM_DEFAULT_ENABLED_RELAY_CHANNELS 0
+#else
+// For SITL, set allowed relay channels to the full mask.
+#define SIM_DEFAULT_ENABLED_RELAY_CHANNELS (1<<16)-1
```
Comment (peterbarker) [inline libraries/SITL/SITL.cpp:56]: You have other options here:
```suggestion
#define SIM_DEFAULT_ENABLED_RELAY_CHANNELS UINT16_MAX
```
or
```suggestion
#define SIM_DEFAULT_ENABLED_RELAY_CHANNELS 0xFFFF
```
or
```suggestion
#define SIM_DEFAULT_ENABLED_RELAY_CHANNELS 0b1111111111111111
```

... all of which I think look a little better than the maths :-)

This'll do, 'though

File: libraries/SITL/SITL.cpp [hunk 4]
```diff
@@ -1029,6 +1038,11 @@ const AP_Param::GroupInfo SIM::var_ins[] = {
     AP_GROUPINFO("GYR5_BIAS",    47, SIM, gyro_bias[4], 0),
 #endif
 
+    // @Param: OH_RELY_MSK
```
Comment (peterbarker) [inline libraries/SITL/SITL.cpp:1041]: I keep reading this as the "Oh, really?" mask.  Just me?

**Review by peterbarker — APPROVED** [review]
LGTM

File: libraries/SITL/SITL.cpp (hunk 4 above)
Comment (joshanne) [reply libraries/SITL/SITL.cpp:1041]: :rofl: You're not the only one. `RELY` is the name for the `AP_Relay` log entry. Though I think I have one letter spare for the parameter name - I could probably make it `RELAY`.

File: libraries/SITL/SITL.cpp (hunk 3 above)
Comment (joshanne) [reply libraries/SITL/SITL.cpp]: Updated to remove maths.

**Comment by peterbarker** [comment]
Merged, thanks!

**Pull request merged**